### ISSUE TITLE
Modules support

### ISSUE THREAD

**Issue by sam-mccall** [clangd/clangd#1293]
It would be nice if some of {C++20 modules, clang header modules} worked in clangd.

Essentially no configurations of this work today: occasionally things may happen to work but there are many reported problems and crashes. These are not just bugs that can be fixed, a design and new infrastructure is needed.

Because clang's module functionality can be enabled by setting driver flags, it is possible/easy for people to end up in this broken and unsupportable state, and wasting (their + our) time debugging it. We should consider failing early and explicitly instead.

**AFAIK nobody has plans/availability to work on this soon.**

Some issues that need to be addressed:
 - which subset of clang's many module features/flags are in scope?
 - explicit modules provided by build system: we can't assume these are compatible (may be a different compiler version). So may need to convert to implicitly building modules needed.
 - having AST workers/background threads spawning module builds independently and coordinating through the module cache is probably not good, we'll want some kind of worker pool and coordination
 - we presumably want to parse modules in the same lite mode as preambles (e.g. skip-function-bodies)
 - gcc modules flags are different than clang modules flags (and MSVC presumably different again), we need to parse them
 - how do modules fit into the dynamic index? what does a preamble with modules in it look like?
 - scattered places where we assume mainfile/preamble dichotomy, reason about #include structure, etc
 - actual handling of import/export AST nodes

**Comment by HighCommander4** [comment]
Some other open bugs we have on file related to modules are #738, #797, #1105.

**Comment by Trass3r** [comment]
This includes `-fmodules`? What is the current state?
With an llvm LLVM_ENABLE_LIBCXX=ON LLVM_ENABLE_MODULES=ON workspace clangd doesn't crash or anything but produces lots of squiggles while the build itself runs through just fine.
Shouldn't clangd rather remove the arguments if it isn't able to handle them properly? (`-fmodules -fmodules-cache-path=... -Xclang -fmodules-local-submodule-visibility -gmodules`)

**Comment by mathstuf** [comment]
(Note: speaking as a CMake developer here with a focus on how modules work with CMake; other build systems have similar challenges, but not necessarily the same ones)

> gcc modules flags are different than clang modules flags (and MSVC presumably different again), we need to parse them

Note that `gcc`'s module flags may indicate an active participant if `-fmodule-mapper=` points to a socket or process execution. That may not be emulatable (I believe `build2` does this, but CMake uses static files written out during the build).

I suspect that for static analysis, a lot more communication between build systems and tooling to know what is going on. The problem is that with explicit modules even `compile_commands.json` is not useful until a build happens because:

- for `gcc`, the `-fmodule-mapper=` argument is generated at build
- for `clang` and `msvc`, modules are communicated via `@rspfile` arguments that don't exist until the build is performed

This isn't even counting the fact that the information presented there isn't directly usable as any referenced BMI files will be in the host compiler, not something the tool necessarily understands.

**Comment by MK-Alias** [comment]
In the age we live in, a programming language is only as good as it runtimes/frameworks and lastly editor support... I mean, who codes these days with just syntax highlighting? C++ modules are - from a user perspective - a very nice addition to the C++ syntax end really modernizes it. However, it is really being held back by editor support.

AFAIK: C++ modules are only supported by VisualStudio (using msvc ixx files) and CLion (using clang cppm files). If clangd would support it, it will significantly improve editor support and with that its usability/popularity.

As for compilers having there own flags/options/file-extensions for modules - which is regretful - just ignore all that and start with supporting the clangs own formats.

Although this is - currently - not a critical functionality, I personally am hoping to see modules supported by clangd soon!

**Comment by sdykae** [comment]
from https://gitlab.kitware.com/cmake/cmake/-/issues/18355
![image](https://user-images.githubusercontent.com/50224521/209905130-73e17c0a-c845-4320-80ae-37e87066a56f.png)
https://github.com/mathstuf/llvm-project/tree/p1689r5-csd
Waiting for stable llvm pr,  and support to clangd to  start development with vscode extension.

**Comment by HighCommander4** [comment]
It may be somewhat instructive to look at CLion's approach to supporting modules, which first appeared in the latest 2022.3 release: https://blog.jetbrains.com/clion/2022/11/clion-2022-3-released/#cpp_modules

It's somewhat light on details but this part seems relevant:

> CLion collects module information from all .ixx, .cppm, and .mxx files in the project and doesn’t rely on the specific approach you use to build modules. Whether it’s CMake with the Visual Studio C++ toolchain or CMake with Clang and specific compilation flags, CLion provides a similar experience

**Comment by ChuanqiXu9** [comment]
>> CLion collects module information from all .ixx, .cppm, and .mxx files in the project and doesn’t rely on the specific approach you use to build modules.

This is not strictly correct since it is allowed to declare a module within a `.cppm` files. But it should be acceptable as the first step.

> It may be somewhat instructive to look at CLion's approach to supporting modules, which first appeared in the latest 2022.3 release: https://blog.jetbrains.com/clion/2022/11/clion-2022-3-released/#cpp_modules

Yeah. Although it didn't mention a lot technical details, I guess we can't learn a lot from this blog. And it only mentioned about the second requirement given by @sam-mccall about the implicit buildings.

----

I guess it may be OK to be too good as the first step. I mean, maybe is it acceptable to ignore some requirements in the first step. Including:
- Compatibility about GCC and MSVC.
- Indexing things.
- lite mode things  (e.g. skip-function-bodies)
- #include structures

And what is good as the first step? I think it may be good if:
- We can collect the potential module files (by its suffixes) in the project tree.
- We can build these module files implicitly in order.
- We can get the AST we just built.

How do you guys think about this? @HighCommander4

**Comment by RedBeard0531** [comment]
> We can collect the potential module files (by its suffixes) in the project tree.

Please don't use extensions to decide which files are modules. It is perfectly valid to have a .cpp file be a module interface. Please just use a scanner. A lot of work went in to the standard wording to ensure that it is possible to very efficiently scan the whole codebase. I know some effort was started in clang-scan-deps for this. It would be great it clangd could reuse and possibly improve this code when doing its scan.

**Comment by ChuanqiXu9** [comment]
> > We can collect the potential module files (by its suffixes) in the project tree.
> 
> Please don't use extensions to decide which files are modules. It is perfectly valid to have a .cpp file be a module interface. Please just use a scanner. A lot of work went in to the standard wording to ensure that it is possible to very efficiently scan the whole codebase. I know some effort was started in clang-scan-deps for this. It would be great it clangd could reuse and possibly improve this code when doing its scan.

Yeah, it is better and more clear. I've edited the post.

**Comment by HighCommander4** [comment]
> And what is good as the first step? I think it may be good if:
> 
>    * We can collect the potential module files by clang-scan-deps in the project tree.
> 
>    * We can build these module files implicitly in order.
> 
>    * We can get the AST we just built.
> 
> How do you guys think about this? @HighCommander4

My perspective as a user is that yes, this level of support for modules would already be useful and valuable. I will defer to @sam-mccall for a maintainer's perspective of what would be accepted as a contribution. (I would imagine that it's fine to have various limitations especially if the support is behind an option/flag.)

A few thoughts/questions:

 * To make the performance acceptable, we'll want to persist BMIs rather than rebuilding them each time we encounter a module. (I think all previous comments assume this, I'm just making it explicit.)
 * To avoid stepping on the project's actual build (which may use a different compiler or a different version of clang), we should persist our BMIs in a clangd-specific place (such as `.cache/clangd`), like you said in the previous comment.
 * Is there an opportunity to reuse this persisted BMI cache between the indexer and the preambles of open files? When you open a source file that imports a module `M`, the BMI for `M` may or may not have already been built by the indexer. If it hasn't, I imagine the AST (preamble) build should build it on the spot (and then the indexer won't need to build it again when it gets around to indexing a source file that imports `M`), but if the indexer has already built M's BMI then I imagine the preamble build can just read it.

**Comment by HighCommander4** [comment]
> (1) When the clangd server initializes, will `BackgroundIndex::changedFilesTask` get called?

Yes, when the compilation database is loaded, it calls [this function](https://searchfox.org/llvm/rev/50dfc9e35d72bf783ebc514ad1e48bd4d0767c5d/clang-tools-extra/clangd/index/Background.cpp#104) with all of the entries in the database, which will call `changedFilesTask()`.

**Comment by ChuanqiXu9** [comment]
> Is there an opportunity to reuse this persisted BMI cache between the indexer and the preambles of open files?

I am not pretty sure that I've understood the idea of the indexer. My current imagination is to let the TUScheduler to hold the ModulesDependencyGraph. And when `TUScheduler::update` is called, check if the file has any modules dependencies, if it has such dependencies and the BMI doesn't exist, don't call `ASTWorker::update` now and assign a task to AsyncRunner to build the BMI and call the `ASTWorker::update` then.

Now it is OK roughly for the a-b test case in https://github.com/llvm/llvm-project/issues/58723.

I feel like the there is only one TUScheduler in one project (is this true?) So the BMIs won't be rebuilt every time we opened a new file.

But I feel I missed some points since I didn't touch the indexer at all. I initially want to work in BackgroundIndex but the log shows it is not reached after I open a new project even if I add `-background-index`. Or is it OK to not make the indexer know about the modules?

**Comment by HighCommander4** [comment]
> I feel like the there is only one TUScheduler in one project (is this true?)

I believe so.

> So the BMIs won't be rebuilt every time we opened a new file.

Note, I was also thinking of persisting the BMIs between clangd invocations.

> is it OK to not make the indexer know about the modules?

The indexer also triggers AST builds (for every file in the project that isn't already indexer), that happens around [here](https://searchfox.org/llvm/rev/50dfc9e35d72bf783ebc514ad1e48bd4d0767c5d/clang-tools-extra/clangd/index/Background.cpp#281). That AST build will encounter `import`s, so I guess it should know about modules as well.

**Comment by ChuanqiXu9** [comment]
> Note, I was also thinking of persisting the BMIs between clangd invocations.

It makes sense. I feel like it is OK

> The indexer also triggers AST builds (for every file in the project that isn't already indexed), that happens around [here](https://searchfox.org/llvm/rev/50dfc9e35d72bf783ebc514ad1e48bd4d0767c5d/clang-tools-extra/clangd/index/Background.cpp#281). That AST build will encounter imports, so I guess it should know about modules as well.

My thought is that it works as long as we rewrote the command line to refer to the new BMIs we built.

**Comment by HighCommander4** [comment]
> > The indexer also triggers AST builds (for every file in the project that isn't already indexed), that happens around [here](https://searchfox.org/llvm/rev/50dfc9e35d72bf783ebc514ad1e48bd4d0767c5d/clang-tools-extra/clangd/index/Background.cpp#281). That AST build will encounter imports, so I guess it should know about modules as well.
> 
> My thought is that it works as long as we rewrote the command line to refer to the new BMIs we built.

I think that would work as part of the solution. Another part of the solution would be to ensure that the indexing of source files only starts after the BMIs have been built.

Presumably, something should proactively drive the building of BMIs (not _just_ when you open a source file that depends on it), so that the indexing of source files eventually gets unblocked. Maybe it makes sense to think of building the BMIs as the "first stage" of indexing, and indexing the source files as the "second stage"?

**Comment by ChuanqiXu9** [comment]
> I think that would work as part of the solution. Another part of the solution would be to ensure that the indexing of source > files only starts after the BMIs have been built.
>
> Presumably, something should proactively drive the building of BMIs (not just when you open a source file that depends on it), so that the indexing of source files eventually gets unblocked. Maybe it makes sense to think of building the BMIs as the "first stage" of indexing, and indexing the source files as the "second stage"?

Do you mean only indexing the source files after we built all the BMIs or after we finish the first stage (since it is possible that some BMI is not buildable)? Sounds like a good idea. It can ease the implementation.

**Comment by sam-mccall** [comment]
A few random thoughts...

(Full disclosure: I did some brainstorming about this a few months ago and couldn't come up with anything good, hopefully y'all are more creative & understand modules better!)

- module building looks kinda like indexing, but i think focusing on *background* indexing is a red herring, and preamble building/indexing is where we should begin. We need to be able to open a file and start using it without any significant fraction of the project having been indexed. This is the single most critical latency path and to hit it today (building preamble)  we take shortcuts which are unacceptable for background indexing (e.g. skip-function-bodies). The
- the project may be so large that we cannot enumerate the modules, let alone touch every file. If we want to rely on clang-scan-deps for small projects, that makes sense, but we should abstract it somehow with a query interface that scales (like CompilationDatabase)
- the big performance win with modules is the ability to reuse modules between invocations e.g. share the overlapping parts of preambles rather than starting from scratch. However editing headers will still force us to rebuild all dependent modules/preambles as modules are tightly coupled to their dependencies. (If you built a.pcm#1 against b.pcm#5 you can't reuse it with b.pcm#6).
- we'll need some new abstractions, which are always hard to get right. TUScheduler is basically "full" in terms of complexity, we can't cram many more responsibilities in there without losing ability to reason about it. Obviously it will have to change, but the bulk of complexity should live somewhere else.
- "modules" can be enabled in (too) many ways, and ultimately the design needs to accommodate as many as possible (there's no room for several implementations!). But we should be clear where we're starting (e.g. only fully-explicit command lines with -fmodule-file=foo=bar using c++20 syntax?) and roughly how other variants will fit in (can we translate implicit to explicit using clang-scan-deps? Do module maps make any difference?)

**Comment by RedBeard0531** [comment]
(note: I know nothing about clangd's internals so this comment is based on guesses based on observable behavior. Hopefully it isn't complete nonsense and if needed someone can't translate into the correct terminology)

I assume clangd must already have an implementation of a proto-buildsystem internally. For example, it can schedule indexing tasks with a controlled amount of parallelism and trigger reindexing of downstream TUs when a header-file changes. Ideally it also notices when files change outside the editor (eg git) or when the editor is closed and is able to do incremental reindexing where needed (although I've experienced issues around this in the past, less so recently). 

I think part of the solution is to double-down on having an integrated build system. Either continue improving the integrated one, or switch to something like llbuild. The key thing that I don't think was required before was task dependencies. While you did have source/header file -> TU indexer job dependencies, they merely triggered rerunning the task. Now you will need some indexer tasks to have an input dependency on the BMI which will be generated by another task. And it is important to rerun the downstream indexers whenever an upstream BMI is meaningfully changed (defining "meaningfully" can be tricky; a first pass can just treat any content difference as meaningful, but I suspect that can be improved on). 

Additionally you cannot begin doing (much of) anything prior to scanning to determine module names and import requirements. And of course, when a file changes, it must be rescanned because it may have added an import, or changed which module/partition it declares itself to be part of. Files can also go from not generating a BMI to generating one eg if you add a `export module foo;` line or change `module blah;` to `module blah:blah;`

I don't know how easy or hard it would be to fit the concept of task dependencies into the existing proto-buildsystem, but at least from an outsider's perspective, that seems like the best (of not only) option. Or maybe I've just spent too much time working on buildsystems, so they feel like the solution to every problem now... 😄

**Comment by sam-mccall** [comment]
> Is there an opportunity to reuse this persisted BMI cache between the indexer and the preambles of open files?

My main concerns here: skip-function-bodies make

**Comment by sam-mccall** [comment]
> I am not sure what you mean for indexing. Abstractly, do you mean a map from source files to the BMI? If yet, I think we're on the same page.

Right - we have to parse the source code and process the results, producing respectively index info or BMI, and we want to do so for roughly the same set of sources, triggered by the same events (when a file is opened: all transitively imported things. In the background: all things in the project).

> Yeah, this is the thing I want to talk to. In my local experiments, I find it is hard/impossible to make modules for clangd work well if we don't do pre-scanning the whole project initially

Agree, neither C++20 modules nor clang header modules don't provide a scalable way to discover module structure. This is a pretty major flaw we need to design around. Though it's basically the same difficulty with resolving `#includes`, and admits some of the same solutions:

 - we can scan the whole project lexically
 - we can have the build system provide an artifact like "modules.json", generated at e.g. cmake time
 - we can query the build system at runtime (e.g. bazel knows module boundaries)

It's fine to fall back to #1, but we need a system that can be extended to #2 or #3.

> And I can image another bad case. Since we'd like to replace -fmodule-file=<module-name>=<path-to-BMI>. But without the pre-scanning for the whole project, we can't know if <module-name> is a third-party module. So that we can't decide to replace it or not.

I'm not sure exactly what you mean by "third-party module", but I don't see a way we can safely consume modules that clangd itself didn't build. (Version skew, among other things). So I think we'll need to be at least rewriting *all* module flags, but ideally programatically replacing all the module loading/caching/resolving interfaces.


module building looks kinda like indexing

I am not sure what you mean for indexing. Abstractly, do you mean a map from source files to the BMI? If yet, I think we're on the same page.

We need to be able to open a file and start using it without any significant fraction of the project having been indexed. This is the single most critical latency path and to hit it today (building preamble) we take shortcuts which are unacceptable for background indexing (e.g. skip-function-bodies).

the project may be so large that we cannot enumerate the modules, let alone touch every file. If we want to rely on clang-scan-deps for small projects, that makes sense, but we should abstract it somehow with a query interface that scales (like CompilationDatabase)

Yeah, this is the thing I want to talk to. In my local experiments, I find it is hard/impossible to make modules for clangd work well if we don't do pre-scanning the whole project initially. In my experiments, currently, I only do on demand modules building and parsing. It may be OK for non-modules code since every TU don't dependent on each other.

But this is problematic with modules. For the simple A-B example,

// b.cppm
export module b;
export int bbbb = 44;

// a.cppm
export module a;
import b;
export int a() {
    return bbbb;
}
Currently, my clangd works if I open b.cppm and a.cppm both. But if I clear the caches and close the 2 tabs and restart the clangd, then when I open a.cppm, I see the annoying red line again. The red line only disappears after I open b.cppm. And this may be super bad in a big project since it makes sense that no one will open every page before he works.

Since now I only scan on demands, when I see import b;, I don't know where module b comes from for real.

And I can image another bad case. Since we'd like to replace -fmodule-file=<module-name>=<path-to-BMI>. But without the pre-scanning for the whole project, we can't know if <module-name> is a third-party module. So that we can't decide to replace it or not.

So now I feel it is impossible to support modules well the scanning initially. Also I agree that it sounds bad. Any ideas?

we'll need some new abstractions, which are always hard to get right. TUScheduler is basically "full" in terms of complexity, we can't cram many more responsibilities in there without losing ability to reason about it. Obviously it will have to change, but the bulk of complexity should live somewhere else.

Agreed. But let's talk about this after we have some real demo.

"modules" can be enabled in (too) many ways, and ultimately the design needs to accommodate as many as possible (there's no room for several implementations!). But we should be clear where we're starting (e.g. only fully-explicit command lines with -fmodule-file=foo=bar using c++20 syntax?) and roughly how other variants will fit in (can we translate implicit to explicit using clang-scan-deps? Do module maps make any difference?)

> We don't have so-many ways to enable C++20 modules. Maybe you have the impression due to the existence of clang modules.

Are you suggesting we only support C++20 named modules? Adding the complexity needed for modules in general but leaving out clang modules seems like a tough sell.
(FWIW my understanding is both Google and Meta are interested in having modules work in clangd and contributing to this, both are using clang modules rather than C++20 modules)

Also, I believe clang modules are at least partly covered by C++20 as "header units". Whether or not we choose not to call them "modules", it seems like we should support codebases that use them.

> For example, we won't have implicit dependencies within C++20 modules.

OK, my understanding is a bit hazy. Which of the following does clang's impl of C++20 modules require/forbid?

 - all imported modules have been built in a separate invocation, before the importing code is built
 - paths to all the imported modules are provided explicitly (rather than being probed for in some cache dir)
 - names of the imported modules are provided explicitly (rather than being retrieved from the PCM file)

> BTW, ideally, the diamond dependency won't be a problem since we will require the consistent build environments.

I mentioned the diamond dependency problem specifically because we currently have two build environments (skip-function-bodies in preamble vs not when background-indexing). I agree that if we keep our build configurations isolated and are content to rebuild all dependent modules when a source file changes, we won't have a diamond dependency problem. (But each of these cost performance)

**Comment by RedBeard0531** [comment]
>> the project may be so large that we cannot enumerate the modules, let alone touch every file. 
>
> So now I feel it is impossible to support modules well the scanning initially. Also I agree that it sounds bad. Any ideas?

I'm not sure how bad it really is. A lot of work went into putting restrictions into the C++20 specification to ensure that it will be possible to write a very fast scanner. It should be possible to scan at or near disk speeds on a single core (maybe not as fast as a high-end NVMe, but certainly several hundred MB/s per core should be achievable). So I think this should scale up to any meaningful definition of "project". This still won't scall up to a 1TB mega-monorepo where you never check out the whole repo at once, and certainly never build it all locally. So you will need an alternative for them (luckily most of them already are willing to invest in custom tooling), but scanning all project files should still be fine for the vast majority of clangd users. I think (admittedly, without evidence yet) a reasonable rule of thumb would be that anything small enough to have a single compile_commands.json file (which implies parsing a single json file with the full command line for every TU) should be able to scan all files within it. @bigcheese might be able to comment more authoritatively on the performance of clang-scan-deps specifically.

That said, you don't have to eagerly do a full scan on everything. You can lazily determine the import DAG while indexing or loading the AST of the file being edited. The only thing you really need to know upfront is the mapping from module/partition name -> source file. While not 100% spec compliant[1], it is probably acceptable to do a "preflight scan" and search all files with the regex `\bmodule\b` and only do a full scan on them since they are the only ones that can define named modules (it is non-obvious, but [this grammar rule](http://eel.is/c++draft/cpp#nt:module-file) means that `module` must appear in the source file itself, not only via an `#include`). Hopefully, in the (far) future the vast majority of C++ files will have `module` in them so this may not make sense forever, but since that is far from true today, you can save some time this way.  Megaprojects, and possibly whatever the equivalent of /usr/include is for modules (still TBD), will need some sort of metadata (static files or server) to find the source for each named module.

 [1] Ideally [line splicing](http://eel.is/c++draft/lex.phases#1.2.sentence-2) wouldn't be legal inside of a keyword or directive. But I think anyone who splits `module` across physical lines outside of a compiler stress test deserves what they get...


> If we're going to have a build system we block on (and modules ~requires it) then it needs to be something new. It's going to be responsible for the critical path to user-facing latency, as opposed to background index's "process an infinite amount of non-critical batch work".

I think "~requires" is a good way to put it. The standard is very carefully worded to avoid placing restrictions on the implementation details. So a build system isn't technically required, and neither are BMIs. In theory, an implementation could parse each TU by simply starting at the top of the file and handling `import` similarly to `#include` by simply reading the source file directly, but in a special context to preserve the modular semantics. I don't know if clang supports such a mode, but I assume that the building blocks would be very useful since they would enable the equivalent of single-file `-E -frewrite-includes` roll-ups of TUs for bug reports and testing. Such a mode (if it exists) may be the shortest path to getting modules working in clangd, since it will at least "work" for both indexing and the open file AST, even if it is slower than it should be.

But having a build system (shorthand for any DAG-walking parallel task runner and incremental re-runner, not restricted to the traditional concept of a build system) and building some sort of BMI (a reusable cache of a module (unit)'s interface) is definitely what the design of C++ modules is optimized for. Hopefully it will result in significantly faster (re)indexing, and open file loading/parsing. Once you have that in place, about the only time I could imagine parsing modules directly from source being useful is when you cold-open a project and want to minimize the latency to first parse of a single open file. But even then, I'm not sure it will actually have lower latency than telling the build system to pause what it is doing and put all resources into building the BMIs for the transitive deps of the current file. Consider that it will be much simpler to parallelize the parsing of independent nodes in the DAG if they are separate tasks, so given enough cores (and core counts are more likely to go up than down over time!) your latency will be proportional to the depth of the DAG from your source file rather than its total size. And this becomes even more of a win if there are multiple open files (eg reopening a project with the same editor views as last time, or just restarting clangd) since they can share the benefit of parsing common deps only once.

> full ASTs are much larger => slower to load

I'm not a clang dev, but from hallway coverversions my understanding is that all compilers are planning to have as close to an O(1) loading cost for BMIs. So the cost of adding an import to a file should be as close to zero as possible, and you will only have to pay for loading the parts of the BMI that you actually use, even if the BMI is huge. I think this will involve both carefully designed on-disk data structures and using mmap to load the BMI and use those data structures in-place. So I wouldn't worry too much about this, at least not yet. Of course I could see a concern about the cost to *generate* the full vs partial BMI contributing to file-open latency, but that is a separate issue. While it may make sense to generate a partial BMI if a full one isn't available yet, it probably isn't worth generating a partial one to save loading time when a full BMI is ready to use. Either way, none of that is on the critical engineering path to getting minimally working module support into clangd, so it may be better to start with only full BMIs and add partials as a possible optimization later. I assume anyone using modules would prefer to get it working sooner, even if it is slower than ideal.

**Comment by ChuanqiXu9** [comment]
> - we can scan the whole project lexically
> - we can have the build system provide an artifact like "modules.json", generated at e.g. cmake time
> - we can query the build system at runtime (e.g. bazel knows module boundaries)
> 
> It's fine to fall back to https://github.com/clangd/clangd/issues/1, but we need a system that can be extended to https://github.com/clangd/clangd/issues/2 or https://github.com/clangd/clangd/issues/3.

I feel like the second point is good. (bazel know nothing about C++20 Named modules now). And it should be good to implement 1 at first.

> I'm not sure exactly what you mean by "third-party module", but I don't see a way we can safely consume modules that clangd itself didn't build. (Version skew, among other things). So I think we'll need to be at least rewriting all module flags, but ideally programatically replacing all the module loading/caching/resolving interfaces.

I mean modules whose source codes are not in the tree of the project. For example, we'll have `import std;` in LLVM17 (in planning), and clearly the source code of `std modules` won't be part of the project. How can we find the source codes? Especially for other third-party modules.

>  but I don't see a way we can safely consume modules that clangd itself didn't build. (Version skew, among other things). 

My thought is that clangd won't need to consume such modules. It just fallbacks to the compiler to consume them. And it's the responsibility of users to make sure such modules are usable. I am not opposite to build modules by clangd as much as possible. I just feel it is not easy to find the corresponding source codes out of tree.

> Are you suggesting we only support C++20 named modules? Adding the complexity needed for modules in general but leaving out clang modules seems like a tough sell.
(FWIW my understanding is both Google and Meta are interested in having modules work in clangd and contributing to this, both are using clang modules rather than C++20 modules)

C++20 Named modules are significantly different from clang header modules and header units. Since named modules is a TU by itself but clang header modules and header units have header's semantics. So I image it may be OK to treat them as headers for clangd. Or at least I think it is OK to consider C++20 Named Modules at first. Since it should be the most complex one.

> OK, my understanding is a bit hazy. Which of the following does clang's impl of C++20 modules require/forbid?
> 
> - all imported modules have been built in a separate invocation, before the importing code is built
> - paths to all the imported modules are provided explicitly (rather than being probed for in some cache dir)
> - names of the imported modules are provided explicitly (rather than being retrieved from the PCM file)

All of the three are required by C++20 named modules.

> I'm not sure how bad it really is. A lot of work went into putting restrictions into the C++20 specification to ensure that it will be possible to write a very fast scanner.

Yeah, now we get the consensus to scan the whole project at the first.

> , it is probably acceptable to do a "preflight scan" and search all files with the regex \bmodule\b and only do a full scan on them since they are the only ones that can define named modules

I am not sure if it will be really faster. Let's re-see this after we realized the full-scanning is bottle-neck.

>  I don't know if clang supports such a mode, but I assume that the building blocks would be very useful since they would enable the equivalent of single-file -E -frewrite-includes roll-ups of TUs for bug reports and testing. Such a mode (if it exists) may be the shortest path to getting modules working in clangd, since it will at least "work" for both indexing and the open file AST, even if it is slower than it should be.

Clang doesn't support this mode now. It looks like there are some extra works to implement it especially to remain the modules semantics. In my brain I feel like it is farer path.

> but from hallway coverversions my understanding is that all compilers are planning to have as close to an O(1) loading cost for BMIs. 

Currently, it is not strictly `O(1)` in clang and I guess we won't never achieve that. But the cost is really low in my experience.

> so it may be better to start with only full BMIs and add partials as a possible optimization later. I assume anyone using modules would prefer to get it working sooner, even if it is slower than ideal.

Strongly agreed : )

**Comment by ChuanqiXu9** [comment]
> we can scan the whole project lexically
> we can have the build system provide an artifact like "modules.json", generated at e.g. cmake time
> we can query the build system at runtime (e.g. bazel knows module boundaries)
>
> It's fine to fall back to https://github.com/clangd/clangd/issues/1, but we need a system that can be extended to https://github.com/clangd/clangd/issues/2 or https://github.com/clangd/clangd/issues/3.

**Comment by HighCommander4** [comment]
> > Is there an opportunity to reuse this persisted BMI cache between the indexer and the preambles of open files?
> 
> The "lite mode" ASTs (skip-function-bodies etc) may not be suitable for indexing, and vice versa (even opportunistically: full ASTs are much larger => slower to load).

My limited understanding of BMIs is that they're similar to PCHs (which is also what our preambles use), whose deserialization is done more or less on demand (e.g. when iterating all `decls()` of a `NamespaceDecl`). This makes me hopeful that a "full" BMI produced by the indexer can be consumed by a preamble build with fairly good performance (and in particular faster than rebuilding that BMI in "lite mode").

> I don't see a way we can safely consume modules that clangd itself didn't build

While so far we have been discussing an "MVP" where clangd builds all the modules itself, in the longer term I'm hoping that we will be able to let users opt into a mode of operation where BMIs are shared with the project's build (which would require the build compiler to be clang and clangd to be version-locked to it).

**Comment by ChuanqiXu9** [comment]
Update: I sent a patch  (ttps://reviews.llvm.org/D153114) for review now. See the review page for details. It is welcome for all people here to give some comment.

**Comment by ChuanqiXu9** [comment]
Update: https://reviews.llvm.org/D153114 is abandoned but the review opinions are pretty helpful. See https://github.com/llvm/llvm-project/pull/66462 for the newest progress.

**Comment by ChuanqiXu9** [comment]
> Any updates on this? This is one of the last things making me hesitate to adopt modules in my projects as I make heavy use of clangd in my workflow.

We have experimental support now. You can try to enable it by `experimental-modules-support` and provide the compilation database.

**Comment by kamrann** [comment]
I tried this yesterday and immediately ran into the issue noted in #1105 (note: no assertion, but the reported error about mismatching compiler flag).

**Comment by ChuanqiXu9** [comment]
> I tried this yesterday and immediately ran into the issue noted in #1105 (note: no assertion, but the reported error about mismatching compiler flag).

Are you meeting the same error? Where is that flag coming from?

**Comment by kamrann** [comment]
According to the linked issue, `clangd` is/was adding that flag automatically. It's not specified in my build, nor in the `compile_commands.json`. 

So it looks as if perhaps whatever was triggering the assert in that issue got addressed at some stage, but the linked patch for the mismatch fix (which I suppose was just saying "a mismatch in this flag need not prevent module load") never got merged?

**Comment by ChuanqiXu9** [comment]
> According to the linked issue, `clangd` is/was adding that flag automatically. It's not specified in my build, nor in the `compile_commands.json`.
> 
> So it looks as if perhaps whatever was triggering the assert in that issue got addressed at some stage, but the linked patch for the mismatch fix (which I suppose was just saying "a mismatch in this flag need not prevent module load") never got merged?

Oh, just noticed that this was the issue 2 years ago. I will do it. 

BTW, in such cases, maybe you can workaround it by adding `-Xclang fno-validate-pch`

**Comment by kamrann** [comment]
> BTW, in such cases, maybe you can workaround it by adding `-Xclang fno-validate-pch`

Good to know, thanks!

**Comment by ChuanqiXu9** [comment]
> According to the linked issue, `clangd` is/was adding that flag automatically. It's not specified in my build, nor in the `compile_commands.json`.
> 
> So it looks as if perhaps whatever was triggering the assert in that issue got addressed at some stage, but the linked patch for the mismatch fix (which I suppose was just saying "a mismatch in this flag need not prevent module load") never got merged?

FYI, done in https://github.com/llvm/llvm-project/commit/d5cdc26f555918e6f292f16ac3800afc97093aac

**Comment by Decodetalkers** [comment]
Seems clangd now only works with the module build of clang, and does not support gcc.. Emm

**Comment by mathstuf** [comment]
> Seems clangd now only works with the module build of clang, and does not support gcc.. Emm

Yes, this makes sense to me as clang cannot read gcc's module files. `clangd` would need to make its own BMIs rather than relying on what the build would already make.

**Comment by onion108** [comment]
> > According to the linked issue, `clangd` is/was adding that flag automatically. It's not specified in my build, nor in the `compile_commands.json`.
> > So it looks as if perhaps whatever was triggering the assert in that issue got addressed at some stage, but the linked patch for the mismatch fix (which I suppose was just saying "a mismatch in this flag need not prevent module load") never got merged?
> 
> FYI, done in [llvm/llvm-project@d5cdc26](https://github.com/llvm/llvm-project/commit/d5cdc26f555918e6f292f16ac3800afc97093aac)

I still encountered this situation (the error of `pch_langopt_mismatch`: retain documentation ...  blahblah) after the commit (I'm using homebrew LLVM 19.1.6, which should contain this commit I guess). Tried to add the `-Xclang -fno-validate-pch` and boom my clangd just died and doesn't give any completion/diagnostics etc.

**Comment by onion108** [comment]
Somehow triggered `invalid AST` when using modules and clangd just stopped working (no completion & no diagnostics & no go-to etc.)
```
I[21:10:51.149] Indexing c++23 standard library in the context of /Users/onion27/Workspace/Programming/Lab/mommy-i-wanna-use-cpp-modules/main.cc
E[21:10:51.156] Failed to scan baseline of /Users/onion27/Workspace/Programming/Lab/mommy-i-wanna-use-cpp-modules/main.cc: failed BeginSourceFile
I[21:10:51.162] BeginSourceFile() failed when building AST for /Users/onion27/Workspace/Programming/Lab/mommy-i-wanna-use-cpp-modules/main.cc
```
Found this in the log, and after that all LSP requests were responded with "invalid AST" and code -32001

**Comment by onion108** [comment]
> Somehow triggered `invalid AST` when using modules and clangd just stopped working (no completion & no diagnostics & no go-to etc.)
> 
> ```
> I[21:10:51.149] Indexing c++23 standard library in the context of /Users/onion27/Workspace/Programming/Lab/mommy-i-wanna-use-cpp-modules/main.cc
> E[21:10:51.156] Failed to scan baseline of /Users/onion27/Workspace/Programming/Lab/mommy-i-wanna-use-cpp-modules/main.cc: failed BeginSourceFile
> I[21:10:51.162] BeginSourceFile() failed when building AST for /Users/onion27/Workspace/Programming/Lab/mommy-i-wanna-use-cpp-modules/main.cc
> ```
> 
> Found this in the log, and after that all LSP requests were responded with "invalid AST" and code -32001

Oh it turned out that it's because I'm too dumb and forgot to update `compile_commands.json` when adding `-Xclang -fretain-comments-from-system-headers`.

Speaking of "-fretain-comments-from-system-headers", it seems that neither `clang --help` nor `clang -cc1 --help` tells me about this option. I waste a lot of time finding what the option is called because of it.

**Comment by ChuanqiXu9** [comment]
> > Somehow triggered `invalid AST` when using modules and clangd just stopped working (no completion & no diagnostics & no go-to etc.)
> > ```
> > I[21:10:51.149] Indexing c++23 standard library in the context of /Users/onion27/Workspace/Programming/Lab/mommy-i-wanna-use-cpp-modules/main.cc
> > E[21:10:51.156] Failed to scan baseline of /Users/onion27/Workspace/Programming/Lab/mommy-i-wanna-use-cpp-modules/main.cc: failed BeginSourceFile
> > I[21:10:51.162] BeginSourceFile() failed when building AST for /Users/onion27/Workspace/Programming/Lab/mommy-i-wanna-use-cpp-modules/main.cc
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Found this in the log, and after that all LSP requests were responded with "invalid AST" and code -32001
> 
> Oh it turned out that it's because I'm too dumb and forgot to update `compile_commands.json` when adding `-Xclang -fretain-comments-from-system-headers`.
> 
> Speaking of "-fretain-comments-from-system-headers", it seems that neither `clang --help` nor `clang -cc1 --help` tells me about this option. I waste a lot of time finding what the option is called because of it.

This might be fixed in https://github.com/llvm/llvm-project/commit/d5cdc26f555918e6f292f16ac3800afc97093aac

**Comment by onion108** [comment]
> > > Somehow triggered `invalid AST` when using modules and clangd just stopped working (no completion & no diagnostics & no go-to etc.)
> > > ```
> > > I[21:10:51.149] Indexing c++23 standard library in the context of /Users/onion27/Workspace/Programming/Lab/mommy-i-wanna-use-cpp-modules/main.cc
> > > E[21:10:51.156] Failed to scan baseline of /Users/onion27/Workspace/Programming/Lab/mommy-i-wanna-use-cpp-modules/main.cc: failed BeginSourceFile
> > > I[21:10:51.162] BeginSourceFile() failed when building AST for /Users/onion27/Workspace/Programming/Lab/mommy-i-wanna-use-cpp-modules/main.cc
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > Found this in the log, and after that all LSP requests were responded with "invalid AST" and code -32001
> > 
> > 
> > Oh it turned out that it's because I'm too dumb and forgot to update `compile_commands.json` when adding `-Xclang -fretain-comments-from-system-headers`.
> > Speaking of "-fretain-comments-from-system-headers", it seems that neither `clang --help` nor `clang -cc1 --help` tells me about this option. I waste a lot of time finding what the option is called because of it.
> 
> This might be fixed in [llvm/llvm-project@d5cdc26](https://github.com/llvm/llvm-project/commit/d5cdc26f555918e6f292f16ac3800afc97093aac)

Yeah it was fixed but what makes me struggle is that the option `-fretain-comments-from-system-headers` doesn't appear in help message of either `clang` or `clang -cc1`, and I put a lot of effort finding that option and adds to my build script.
<img width="1440" alt="スクリーンショット 2024-12-28 14 08 13" src="https://github.com/user-attachments/assets/429a5272-f04c-499a-8e03-fa46df1f5bd7" />
(When I'm typing "/retain documentation comments" trying to search in the help but nothing matches. The option does exist on my compiler though because I put `-fretain-comments-from-system-headers` and it just worked. But it's still a little bit confusing 'cause when I can't search anything about the flag in the help I almost think that I have the wrong version of the compiler, but it turned out that the option actually exists but somehow doesn't appear in the help)

**Comment by YanzhaoW** [comment]
Hi, 

I have tried the module support from clangd and it kind of "works" with some problems/limitations:

1. If a new symbol is added in another file, clangd seems to only able to produce correct diagnosis after the compilation of the whole project.

   This could beat the very purpose of using clangd in the first place: **to see the compilation errors on the fly without compiling the whole project.** Now because of the modules, the new type information is stored in the BMI files which are only available during the compilation. I need to compile the project first to get the diagnosis from clangd, which kind of loses its purpose as I already have the diagnosis from the compilation.

2. Can't `rename` symbols because of "Error on prepareRename: Cannot rename symbol: symbol may be used in other files (not eligible for indexing)"

**Comment by mathstuf** [comment]
Note that `clangd` would need to learn to compile its own BMIs for a given build anyways to support projects using, e.g., `gcc`. The existing `compile_commands.json` just does not have enough information for this. See [my CppCon 2024 talk](https://www.youtube.com/watch?v=GUqs_CM7K_0) for details.

**Comment by ChuanqiXu9** [comment]
> Hi,
> 
> I have tried the module support from clangd and it kind of "works" with some problems/limitations:
> 
> 1. If a new symbol is added in another file, clangd seems to only able to produce correct diagnosis after the compilation of the whole project.
>    This could beat the very purpose of using clangd in the first place: **to see the compilation errors on the fly without compiling the whole project.** Now because of the modules, the new type information is stored in the BMI files which are only available during the compilation. I need to compile the project first to get the diagnosis from clangd, which kind of loses its purpose as I already have the diagnosis from the compilation.
> 2. Can't `rename` symbols because of "Error on prepareRename: Cannot rename symbol: symbol may be used in other files (not eligible for indexing)"

Have you used the latest clangd and enable experiemental modules support?

**Comment by ChuanqiXu9** [comment]
> Note that `clangd` would need to learn to compile its own BMIs for a given build anyways to support projects using, e.g., `gcc`. The existing `compile_commands.json` just does not have enough information for this. See [my CppCon 2024 talk](https://www.youtube.com/watch?v=GUqs_CM7K_0) for details.

Yeah, I met the problem in our downstream that we have multiple module units declare the same "module". Maybe you can open an issue in llvm

**Comment by chriselrod** [comment]
> Have you used the latest clangd and enable experiemental modules support?

I have been using llvm 21-rc2 with `--experimental-modules-support`, but still need to recompile to get updated BMIs. Is there something else I need to do, or does this require building a newer llvm (21.1 or 22)?

**Comment by YanzhaoW** [comment]
> Have you used the latest clangd and enable experiemental modules support?

I'm using clangd `20.1.7` (having `--experimental-modules-support` enabled), which is the same version as clang's. If I use latest `20.1.8`, it won't work because of "clang: Module file built from a different branch than the compiler"

**Comment by ChuanqiXu9** [comment]
> > Have you used the latest clangd and enable experiemental modules support?
> 
> I have been using llvm 21-rc2 with `--experimental-modules-support`, but still need to recompile to get updated BMIs. Is there something else I need to do, or does this require building a newer llvm (21.1 or 22)?

> > Have you used the latest clangd and enable experiemental modules support?
> 
> I'm using clangd `20.1.7` (having `--experimental-modules-support` enabled), which is the same version as clang's. If I use latest `20.1.8`, it won't work because of "clang: Module file built from a different branch than the compiler"

Please try the latest clangd in trunk (maybe you need to compile clangd by yourself or maybe you can find prebuilt packages). And if it fails, you can see the logs by enabling `--log=verbose` and `--debug-modules-builder` to see the logs and you can try to see more logs .

**Comment by chriselrod** [comment]
Unrelated, but have you seen a ton of `reference-tu-local-entity-in-other-tu` warnings with modules on clang trunk? I didn't get these with LLVM 21, but did after building trunk.
Any advice on addressing these? I tried adding explicit instantiations, but I couldn't find the correct function-argument signatures for some (there functions were overloaded, with several definitions having the same template parameters). These functions are calling `static` functions from the immintrin header, so perhaps that's the cause (although it points to definitions within the module). Would be nice if it gave the full list of argument types (that you can mostly copy/paste into an explicit instantiation), or at least a line number.

I'll try more tonight to see if this lets it recompile modules, but I saw multiple-TU errors with first look. Don't read too much into that until I report back with more information later.

**Comment by ChuanqiXu9** [comment]
> Unrelated, but have you seen a ton of `reference-tu-local-entity-in-other-tu` warnings with modules on clang trunk? I didn't get these with LLVM 21, but did after building trunk. Any advice on addressing these? I tried adding explicit instantiations, but I couldn't find the correct function-argument signatures for some (there functions were overloaded, with several definitions having the same template parameters). These functions are calling `static` functions from the immintrin header, so perhaps that's the cause (although it points to definitions within the module). Would be nice if it gave the full list of argument types (that you can mostly copy/paste into an explicit instantiation), or at least a line number.
> 
> I'll try more tonight to see if this lets it recompile modules, but I saw multiple-TU errors with first look. Don't read too much into that until I report back with more information later.

Since we implement more checks in clang22. This is required by standards. The spec doesn't allow other TU to reference TUlocal in other TU.

**Comment by chriselrod** [comment]
I suspect the check is yielding false positives, but do not have a minimal reproducer.

**Comment by kamrann** [comment]
There were a few false positives when GCC first implemented this so it may well be the case, but there were far, far more genuine cases. The reality is there is heaps of code out there (numerous cases in boost for example) that is ill-formed by these new rules, most of which are questionable even in a non-modules case. The `static` functions in system headers being a clear example.

If you're in control of the code in the module interface, you can work around things like this by wrapping the problematic elements, for example don't reference included static functions inside a templated/inline function, and instead create a non-exported non-inline function that calls the included one, then use that within the exported entity. Of course, it can potentially be a bunch of work.

**Comment by RedBeard0531** [comment]
> The static functions in system headers being a clear example.

That was _supposed_ to be fixed by http://wg21.link/P2003 (which was approved for C++20 via a merge into http://wg21.link/P1815). But it looks like it only solved the issue for header units, not headers textually included into the GMF. This feels more like a specification bug than design intent.

\<opinion>
This was probably because we were assuming a world where once your compiler+build system supported modules you would include basically all "real headers" as header units rather than textually. As with many of the issues with modules, I think we would be in a much better place if that were the case. For example, there would be no issue merging decls from `import std;` with `#include <vector>` since the former would already have made all of the decls from the latter reachable and the `#include` (which would automatically be treated as an `import`) would only pull in the macros and make some declarations that were already _reachable_ also be _visible_.
\</opinion>

**Comment by chriselrod** [comment]
> I suspect the check is yielding false positives, but do not have a minimal reproducer.

```
In file included from /home/chriselrod/Documents/progwork/cxx/LoopModels/extern/Math/mod/Ma
th/Array.cxxm:10:
/home/chriselrod/Documents/progwork/cxx/LoopModels/extern/Math/mod/Math/ArrayOps.cxxm:465:4
6: error: instantiation of 'operator=' triggers reference to TU-local entity 'store<long>' 
from other TU 'SIMD' [-Werror,-Wreference-tu-local-entity-in-other-tu]
  465 |           [=](auto &d, const auto &s) { d[u] = s; });
      |                                              ^
1 error generated.
```
`operator=` is a template function within the SIMD interface module. 
It is calling a template function `store<long>`, which in turn calls some `static` function from the `immintrin.h` header, which was included in the GMF.

> There were a few false positives when GCC first implemented this so it may well be the case, but there were far, far more genuine cases. The reality is there is heaps of code out there (numerous cases in boost for example) that is ill-formed by these new rules, most of which are questionable even in a non-modules case. The static functions in system headers being a clear example.
> 
> If you're in control of the code in the module interface, you can work around things like this by wrapping the problematic elements, for example don't reference included static functions inside a templated/inline function, and instead create a non-exported non-inline function that calls the included one, then use that within the exported entity. Of course, it can potentially be a bunch of work.

In this case `store` is a pretty trivial function, as is `operator=`.
It'd be easy enough to make `store<T>` (for example) non-template and enumerate the methods rather than using `if constexpr` branching, and then avoid exporting it -- other modules should be consuming it through the exported `operator=`.

Such trivial functions as `store<T>` should be `inline`, and shouldn't produce any externally visible symbols.
Would I get that from not exporting it, i.e. by not exporting it does it have internal linkage and thus the same behavior w/ respect to optimization?

Also, funny that you bring up `boost`.
My SIMD module does two things:
1. Wraps a bunch of SIMD intrinsics to allow for bulk-SIMD operations consumed by other modules.
2. Includes a few boost headers (related to unordered map) in the GMF, because having the SIMD module also be my boost-undordered module was the only way to avoid clang (not the linker) throwing an error on duplicate definitions of a static function defined in `immintrin.h`, which boost also includes. Making the function no longer static also worked, but patching system headers didn't sound like a real solution. 

Do these functions need to be static? Can we PR clang and glibc to make them all `inline` with hidden visibility?

> That was *supposed* to be fixed by...

Unfortunately, no `import immintrin;`. While C++26 is adding a SIMD library, neither it nor documented vector builtins support all the operations I'm using, so I'm stuck with `#include <immintrin.h>` for now.

**Comment by kamrann** [comment]
> This was probably because we were assuming a world where once your compiler+build system supported modules you would include basically all "real headers" as header units rather than textually. As with many of the issues with modules, I think we would be in a much better place if that were the case. 

This absolutely fits with my experience with modules. I've been using them for a couple of years now but with no end of issues. Mostly due to build tool limitations I haven't used header units at all, and I can say without doubt that a large proportion of the issues I've had stem one way or another from includes in the GMF.

> Do these functions need to be static?

Generally no, but any kind of changes to system headers are probably going to be hard to push through. Also I wasn't aware until reading the above linked PR of the C interop thing, sounds like in some cases the `static inline` combination is intentional. This is going off on a very non-clangd tangent though; I'd suggest opening an issue on LLVM if you suspect something is actually a bug.

**Comment by ChuanqiXu9** [comment]
> > I suspect the check is yielding false positives, but do not have a minimal reproducer.
> 
> ```
> In file included from /home/chriselrod/Documents/progwork/cxx/LoopModels/extern/Math/mod/Ma
> th/Array.cxxm:10:
> /home/chriselrod/Documents/progwork/cxx/LoopModels/extern/Math/mod/Math/ArrayOps.cxxm:465:4
> 6: error: instantiation of 'operator=' triggers reference to TU-local entity 'store<long>' 
> from other TU 'SIMD' [-Werror,-Wreference-tu-local-entity-in-other-tu]
>   465 |           [=](auto &d, const auto &s) { d[u] = s; });
>       |                                              ^
> 1 error generated.
> ```
> 
> `operator=` is a template function within the SIMD interface module. It is calling a template function `store<long>`, which in turn calls some `static` function from the `immintrin.h` header, which was included in the GMF.

It seems not a false positive technically. But give what @RedBeard0531  said, I don't feel bad to make  a whitelist for the static functions in GMF.

> 
> > There were a few false positives when GCC first implemented this so it may well be the case, but there were far, far more genuine cases. The reality is there is heaps of code out there (numerous cases in boost for example) that is ill-formed by these new rules, most of which are questionable even in a non-modules case. The static functions in system headers being a clear example.
> > If you're in control of the code in the module interface, you can work around things like this by wrapping the problematic elements, for example don't reference included static functions inside a templated/inline function, and instead create a non-exported non-inline function that calls the included one, then use that within the exported entity. Of course, it can potentially be a bunch of work.
> 
> In this case `store` is a pretty trivial function, as is `operator=`. It'd be easy enough to make `store<T>` (for example) non-template and enumerate the methods rather than using `if constexpr` branching, and then avoid exporting it -- other modules should be consuming it through the exported `operator=`.
> 
> Such trivial functions as `store<T>` should be `inline`, and shouldn't produce any externally visible symbols. Would I get that from not exporting it, i.e. by not exporting it does it have internal linkage and thus the same behavior w/ respect to optimization?
> 
> Also, funny that you bring up `boost`. My SIMD module does two things:
> 
> 1. Wraps a bunch of SIMD intrinsics to allow for bulk-SIMD operations consumed by other modules.
> 2. Includes a few boost headers (related to unordered map) in the GMF, because having the SIMD module also be my boost-undordered module was the only way to avoid clang (not the linker) throwing an error on duplicate definitions of a static function defined in `immintrin.h`, which boost also includes. Making the function no longer static also worked, but patching system headers didn't sound like a real solution.
> 
> Do these functions need to be static? Can we PR clang and glibc to make them all `inline` with hidden visibility?

Yeah, I think so.

> 
> > That was _supposed_ to be fixed by...
> 
> Unfortunately, no `import immintrin;`. While C++26 is adding a SIMD library, neither it nor documented vector builtins support all the operations I'm using, so I'm stuck with `#include <immintrin.h>` for now.

**Comment by mwasplund** [comment]
@ChuanqiXu9 Has there been any progress on having the build system define the modules database up front or on demand instead of scanning the entire directory structure up front? I would like to experiment with integrating my build system into clangd with full module support and the current design does not play nicely with external package references.

> * we can scan the whole project lexically
> * we can have the build system provide an artifact like "modules.json", generated at e.g. cmake time
> * we can query the build system at runtime (e.g. bazel knows module boundaries)

Most up to date info I could find:
https://chuanqixu9.github.io/c++/2025/12/03/Clangd-support-for-Modules.en.html

**Comment by mathstuf** [comment]
FD: CMake developer

@mwasplund CMake has an experimental feature[^1] to generate "build databases" (see my CppCon talk: https://www.youtube.com/watch?v=GUqs_CM7K_0) which aim to provide the appropriate amount of information for tools like `clangd` to be able to deal with a build tree containing modules. There is interest in implementing support in `clangd`, but not yet a timeline for an implementation. Would you be interested in investigating support for generating build databases from your side and potentially collaborating on `clangd` support as well?

[^1]: This means that it is not subject to CMake's compatibility guarantees as yet and we can evolve it as necessary if we discover gaps while gathering ecosystem-wide implementation experience with them.

**Comment by mwasplund** [comment]
@mathstuf I am interested in helping out. I will find some time to add the ability to convert my internal build database into the schema you are proposing here. If I get to it before anyone else, I can prototype something in clangd for loading the build_database as a replacement for the existing directory scanner.

In my mind this still feels like the wrong direction. Instead of the build system telling clangd the module dependency and compilation arguments, would it not make more sense for clangd to ask the build system for this information? The build system should already have this internal state tracking incremental builds and will have the most up to date information as the project structure evolves during development. This was less of a concern in the good old days of headers where the compile_commands were relatively static during development. However, now any time you change an import the dependency structures need to update which must come from the build system.

**Comment by mathstuf** [comment]
The build system is going to have it in the format that the compiler in use wants. So if `clangd` is going to read `cl.exe` or `gcc`'s `-fmodule-mapper=` files, the build system can certainly provide them.

Also, there is a difference between "build system" and "build tool". `ninja` has no idea about modules; it deals with commands, paths, and dependencies between them. CMake does know about modules, but doesn't really have the whole picture during the build (it only provides itself enough information to deal with specific parts to guarantee its semantics when it needs to).

One detail is that tools may need different BMI compatibility sets than the compiler in use. For example, `clang` might need different BMIs because of some flag that GCC doesn't, so the build graph is actually not useful because the real build wants to share BMIs between two libraries that `clangd` cannot. That is the kind of thing that build databases are supposed to help provide additional context over compilation databases.

**Comment by mwasplund** [comment]
My understanding may be wrong, but I thought clangd would always consume the clang "flavor" of the build definition. So even on project targeting GCC or MSVC it will require clang flags to compile and generate the symbols for the current TU. Is that correct? In this case the build system would have to compile the required BMIs and maintain the dependency and compilation arguments for clang in parallel to the target configuration if they differ. 

In the approach I took for my system I do have the full context of the build graph and the current state of the execution (after phase 1 which performs the scan operations). In such a system we would only require a second invocation for generation with clang configs that are compatible with the instance of clangd. As long as this parallel state is keeps in sync the build system can function as the source of truth for the build database to react to user edits.

**Comment by mathstuf** [comment]
Anything reading `compile_commands.json` needs to understand (or ignore at its own peril) flags it doesn't understand when ingesting into whatever internal machinery is used for the tooling. Build databases are the same way. I can't imagine having to wait for build tools to be upgraded to deploy new tooling across a set of projects; all they can do is write the metadata in a way that is generally understood and then use that as the interoperability point (keeping implementations `O(2N)` instead of `O(N²)`).

As for CMake itself, it is unlikely never going to be able to generate build configurations for multiple compilers due to the imperative way projects can add flags or change the build graph based on compiler information (not to mention the assumptions baked in 25+ years ago that there is a single toolchain for a given language).

**Comment by ChuanqiXu9** [comment]
> [@ChuanqiXu9](https://github.com/ChuanqiXu9) Has there been any progress on having the build system define the modules database up front or on demand instead of scanning the entire directory structure up front? I would like to experiment with integrating my build system into clangd with full module support and the current design does not play nicely with external package references.
> 
> > * we can scan the whole project lexically
> > * we can have the build system provide an artifact like "modules.json", generated at e.g. cmake time
> > * we can query the build system at runtime (e.g. bazel knows module boundaries)
> 
> Most up to date info I could find: https://chuanqixu9.github.io/c++/2025/12/03/Clangd-support-for-Modules.en.html

No update for new module database in clangd. But I tried to improve the ability to dig information in current compilation database. If you're interested, you can try.

**Comment by rakibdev** [comment]
time to rewrite in rust

**Comment by w0rp** [comment]
> time to rewrite in rust

Very funny, but no thanks. 😆

**Comment by Decodetalkers** [comment]
I found that cmake can already generate something like CXXModules.json, Can we use that?

<img width="956" height="1142" alt="Image" src="https://github.com/user-attachments/assets/0e2bbd62-30b5-4699-970c-db0424f21a70" />


we also have this one

<img width="1137" height="1257" alt="Image" src="https://github.com/user-attachments/assets/709255c9-794f-47cf-9d3a-253a00666308" />

**Comment by mathstuf** [comment]
You mean the file CMake writes for itself under `<tgt>.dir/`? That is meant for the collator to do its job; it will still have the "talks about GCC when using GCC as the compiler" problem `clangd` needs to contend with.

**Comment by Decodetalkers** [comment]
@ChuanqiXu9 I found the code that reserved for gcc to run the command, but I do not know how to use it...

I am still trying to make clear about the project of clangd, can you tell me the place when clangd starts scanning the project, and 

<img width="1266" height="181" alt="Image" src="https://github.com/user-attachments/assets/5df5967f-0f36-45dd-9efa-abd7b74ad84d" />

where this error comes from. It seems it is the error by clang++, but when I am using gcc, it runs the first command in compile-commands.json, that means , it runs gcc, and the error is by gcc. or clang++? I do not know. And why do we have to use the pcm files to get the location when we can get it from the file?

```
  "command": "/usr/bin/c++ -std=c++23 -fmodules-ts -fmodule-mapper=CMakeFiles/foo.dir/hello.cxx.o.modmap -MD -fdeps-format=p1689r5 -x c++ -o CMakeFiles/foo.dir/hello.cxx.o -c /home/cht/cpp/modulecpp/hello.cxx",
```

For example, and cmake has already prepared the right location for module file. Can we use that instead? That will be fine for gcc and clang++

**Comment by Decodetalkers** [comment]
> You mean the file CMake writes for itself under `<tgt>.dir/`? That is meant for the collator to do its job; it will still have the "talks about GCC when using GCC as the compiler" problem `clangd` needs to contend with.

I see, but cannot but cannot we skip the clang++ and use the file location directly?

**Comment by Decodetalkers** [comment]
> [@ChuanqiXu9](https://github.com/ChuanqiXu9) I found the code that reserved for gcc to run the command, but I do not know how to use it...
> 
> I am still trying to make clear about the project of clangd, can you tell me the place when clangd starts scanning the project, and
> <img alt="Image" width="1266" height="181" src="https://private-user-images.githubusercontent.com/60290287/596127759-5df5967f-0f36-45dd-9efa-abd7b74ad84d.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NzkzNzQxODcsIm5iZiI6MTc3OTM3Mzg4NywicGF0aCI6Ii82MDI5MDI4Ny81OTYxMjc3NTktNWRmNTk2N2YtMGYzNi00NWRkLTllZmEtYWJkN2I3NGFkODRkLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNjA1MjElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjYwNTIxVDE0MzEyN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTFjZDIwMWZiOWMwZmZmZjBlZmVhOGFlOGUzYzBiZTFiMTgzY2RhMzE2OWI3MWUzNjRjMzNkOTRiNzkzOTEwNTEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JnJlc3BvbnNlLWNvbnRlbnQtdHlwZT1pbWFnZSUyRnBuZyJ9.k_jPNLDNS7pgv4F4QlyGSy01-9Y3SpoG8dkPY3uPOHc">
> 
> where this error comes from. It seems it is the error by clang++, but when I am using gcc, it runs the first command in compile-commands.json, that means , it runs gcc, and the error is by gcc. or clang++? I do not know. And why do we have to use the pcm files to get the location when we can get it from the file?
> 
> ```
>   "command": "/usr/bin/c++ -std=c++23 -fmodules-ts -fmodule-mapper=CMakeFiles/foo.dir/hello.cxx.o.modmap -MD -fdeps-format=p1689r5 -x c++ -o CMakeFiles/foo.dir/hello.cxx.o -c /home/cht/cpp/modulecpp/hello.cxx",
> ```
> 
> For example, and cmake has already prepared the right location for module file. Can we use that instead? That will be fine for gcc and clang++

I have a problem. From the code, clangd should use the first argument in command as the compiler, but seems it always uses the clang++, I want to figure it out. if it is using gcc, I should never get this warning

**Comment by HighCommander4** [comment]
> clangd should use the first argument in command as the compiler, but seems it always uses the clang++, I want to figure it out. if it is using gcc, I should never get this warning

clangd uses _knowledge_ of the first argument to determine things like what system include directories to use, and what compatibility mode to use to interpret the command line arguments, but as a tool built on the clang frontend, the actual processing of the code necessarily happens using the clang frontend, and therefore is subject to the behaviours / supported features / diagnostics of the clang frontend.

**Comment by ChuanqiXu9** [comment]
@Decodetalkers yeah, I think clangd has a feature called command mangler can handle gcc's command which clang can't handle.

But as clangd is built on clang, using clang as the base compiler would give you best experience. And I suggest you to find a nightly built clangd/clang package or trying to build it yourself as there are some improvements in the trunk.

**Comment by Decodetalkers** [comment]
> [@Decodetalkers](https://github.com/Decodetalkers) yeah, I think clangd has a feature called command mangler can handle gcc's command which clang can't handle.
> 
> But as clangd is built on clang, using clang as the base compiler would give you best experience. And I suggest you to find a nightly built clangd/clang package or trying to build it yourself as there are some improvements in the trunk.

Thanks for your reply. But the problem is that I want to make clangd works on gcc modules. That is also what a lsp should do I think. And then how to use command mangler? Is there a document?

**Comment by ChuanqiXu9** [comment]
I don't use gcc so I don't use command mangler too. I can only find https://clang.llvm.org/extra/doxygen/structclang_1_1clangd_1_1CommandMangler.html . @HighCommander4 do you have a document for end users?

**Comment by HighCommander4** [comment]
> clangd has a feature called command mangler can handle gcc's command which clang can't handle

What clangd's command mangler does is:

 * Add `-isystem` flags for system includes extracted from the compiler (if `--query-driver` is being used)
 * Modify the compile command by applying any edits specified in clangd's config file via the [`CompileFlags`](https://clangd.llvm.org/config.html#compileflags) key. This can be used to e.g. remove gcc-specific flags that the clang driver doesn't understand.

There is no general ability to handle gcc commands which clang can't handle beyond this.

> I want to make clangd works on gcc modules

If by that you mean consuming the PCM files produced by gcc, then this is not possible, since gcc's and clang's PCM formats are completely different. Consuming PCM files produced by gcc would require a language server built on gcc's frontend.

(For the record, I think having a language server built on gcc's frontend would be healthy for the C++ community. But that's something to ask about / advocate for in gcc developer channels.)

**Comment by Decodetalkers** [comment]
<img width="859" height="419" alt="Image" src="https://github.com/user-attachments/assets/1930af60-f2fc-4518-b590-5fd55321fa83" />


<img width="696" height="418" alt="Image" src="https://github.com/user-attachments/assets/f13e4166-4f13-4a4b-ab57-2e21ebb8e8f8" />

ok, gcc use readref to get the source 

https://splichal.eu/gccsphinx-final/html/gcc/gcc-command-options/c%2B%2B-modules.html#module-mapper

Emm. it relies on readelf to read the information.. and there is not dependency map in CXXDependency.json or CXXModules.json, because it did not tell who is compiled into who.. I think maybe clangd should not invoke the readelf itself, or inducing the libelf, maybe it will be better be done by cmake.. I do not know.. But I finally know how to get the information from gcm

I can get the source map from g++ modmap, or from CXXModules.json. And use readelf to get the information,

**Comment by mathstuf** [comment]
Please do not dig into CMake internal files for this information (we provide no stability nor compatibility guarantees about these files); build databases should be the way to communicate the structure of the builds for tooling.

**Comment by Decodetalkers** [comment]
Ok.. Sorry for bother. Ok, I will stop digging to cmake file, sorry.

After reading the document of gcc, I have made an idea. The core logic need to modified is

https://github.com/llvm/llvm-project/blob/main/clang-tools-extra/clangd/ProjectModules.cpp#L212-L262

And in the logic of clangd, the command line in compile_commands.json  will be added an information about compiler, named "driver-mode", I can check if it is g++, then switch to another logic.

Then the core logic for gcc is that

First, clangd will read the command argument , for example "-fmodule-mapper=CMakeFiles/foo.dir/hello.cxx.o.modmap",

And use the "direcctory" path to make an absolute path. Then read the modmap

```
$root .
hellob CMakeFiles/foo.dir/hellob.gcm
beta CMakeFiles/foo.dir/beta.gcm
```

The modmap of gcc is like above. Then we can get the depended modules and their gcm. 

After that, I will use readref to read the gcm, and get the data, and check if the source file is the same with the module. If they are the same, then we know this cpp file is a module file and give the result  Result.ModuleName, and its required modules, else, just add the required modules.. This is my idea

**Comment by mwasplund** [comment]
I have a working prototype for build database loading and project modules handling. It is rough, but I hope to send a PR soon to get directional feedback before I polish the code. @mathstuf I had a question I sent to reddit (wanted to discuss without randomizing the thread, let me know if there is a better place to chat quick) and an issue on CMake Discourse which is preventing me from testing the updates to clangd with CMake generated build_database files.

**Comment by mathstuf** [comment]
I responded on Discourse with a workaround for today and an MR for the future :) .

For chatting…email is probably best.

**Comment by Decodetalkers** [comment]
I am trying to add a min support for gcc, I have understand the way to get the dependency tree of gcc modules, and I implemented a scan rule for gcc, The logic is here https://github.com/Decodetalkers/llvm-project/commit/2e37c3823a082be2822ae1130136836b36ae2357#diff-9591a0884bbee5eb3027471c0efdcb90e8c4f63dcb116c0db9a0bf566b3f25f2R219-R403

But I still cannot let the completion works.. @ChuanqiXu9 Sorry for bothering... But I want to do something. Can I get your help? Thanks. Thank you for your blog

**Comment by ChuanqiXu9** [comment]
You can file a new pr and we can discuss there. But I am not sure why do we need to support gcc specially in clangd. We need to define the problem first.

**Comment by Decodetalkers** [comment]
> You can file a new pr and we can discuss there. But I am not sure why do we need to support gcc specially in clangd. We need to define the problem first.

Emm. Now I know what should I do next. I just want to use gcc to build the project and the completion still work there. You know, on linux we almost prefer gcc to clang, and I think making clangd can also complete under the project built by gcc will be helpful. I think I am wrong above, I have already understand how clangd works. Before I just thought if I can got the dependency map, it will work, but still, it needs to be compiled by clang frontend once, then it can get the completions. The problem seems just be the commands in compile_commands.json cannot be used by clang++, but it seems can be convert to that one that clang++ can use. I will take a try. I think as soon as I can get the dependency map, I can let clang frontend also work under with the gcc compile_commands.json, seems what is needed is a simple translation..

**Comment by ChuanqiXu9** [comment]
> > You can file a new pr and we can discuss there. But I am not sure why do we need to support gcc specially in clangd. We need to define the problem first.
> 
> Emm. Now I know what should I do next. I just want to use gcc to build the project and the completion still work there. You know, on linux we almost prefer gcc to clang, and I think making clangd can also complete under the project built by gcc will be helpful. I think I am wrong above, I have already understand how clangd works. Before I just thought if I can got the dependency map, it will work, but still, it needs to be compiled by clang frontend once, then it can get the completions. The problem seems just be the commands in compile_commands.json cannot be used by clang++, but it seems can be convert to that one that clang++ can use. I will take a try. I think as soon as I can get the dependency map, I can let clang frontend also work under with the gcc compile_commands.json, seems what is needed is a simple translation..

Yeah, clangd must use clang to compile the source to start clangd's work.

**Comment by Decodetalkers** [comment]
solved for gcc module clangd support 

https://github.com/llvm/llvm-project/pull/200001 

Can anyone take a try? I think I have fixed it

**Comment by ChuanqiXu9** [comment]
> llvm/llvm-project#200001

I think this is what command mangler does.
